core/src/state/binding.rs

```diff
@@ -1,141 +1,264 @@
+                let store = if store_req.can_deduplicate {
+                    model_data_store.lenses_dedup.entry(store_req.store_id).or_insert_with(|| {
+                        bindable.make_store(model_data).expect(
+                            "Bug: bindable failed to produce a store. Probably an internal error.",
+                        )
+                    })
+                } else {
+                    let idx = model_data_store.lenses_dup.len();
+                    model_data_store.lenses_dup.push(bindable.make_store(model_data).expect(
+                        "Bug: bindable failed to produce a store. Probably an internal error.",
+                    ));
+                    model_data_store.lenses_dup.get_mut(idx).unwrap()
+                };
+
+                // ...so that we can register the observer with the appropriate store.
+                bindable.add_to_store(store.as_mut(), id);
+            }
+        }
+    }
+
+    id
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct StoreId {
+    pub can_deduplicate: bool,
+    pub store_id: TypeId,
+    pub model_id: TypeId,
+}
+
+pub trait Bindable: 'static + Clone {
+    type Output;
+    fn view<D: DataContext, F: FnOnce(Option<&Self::Output>) -> T, T>(
+        &self,
+        cx: &D,
+        viewer: F,
+    ) -> T;
+    fn get<D: DataContext>(&self, cx: &D) -> Option<Self::Output>
     where
-        F: 'static + Fn(&mut Context, L),
+        Self::Output: Clone,
     {
-        let binding = Self { lens: lens.clone(), content: Some(Box::new(builder)) };
-
-        let id = cx.entity_manager.create();
-        let current = cx.current();
-        cx.tree().add(id, current).expect("Failed to add to tree");
-        cx.cache().add(id).expect("Failed to add to cache");
-        cx.style().add(id);
-
-        let ancestors = cx.current().parent_iter(cx.tree()).collect::<HashSet<_>>();
-        let new_ancestors = id.parent_iter(cx.tree()).collect::<Vec<_>>();
-
-        for entity in new_ancestors {
-            if let Some(model_data_store) = cx.data.get_mut(entity) {
-                if let Some(model_data) = model_data_store.data.get(&TypeId::of::<L::Source>()) {
-                    if let Some(lens_wrap) =
-                        lens.cache_key().and_then(|key| model_data_store.lenses_dedup.get_mut(&key))
-                    {
-                        let observers = lens_wrap.observers();
-
-                        if ancestors.intersection(observers).next().is_none() {
-                            lens_wrap.add_observer(id);
-                        }
-                    } else {
-                        let mut observers = HashSet::new();
-                        observers.insert(id);
+        self.view(cx, |v| v.cloned())
+    }
 
-                        let model = model_data.downcast_ref::<L::Source>().unwrap();
+    fn requests(&self) -> Vec<StoreId>;
+    fn make_store(&self, source: ModelOrView) -> Option<Box<dyn StoreHandler>>;
```

Comment:
Could it be worth using an associated type for the returned store?

Reply:
I considered this, but I'd like to keep the flexibility of a single bindable being able to bind to multiple kinds of stores!

Reply:
But how would a single bindable choose what store to use?

Reply:
The thing I was envisioning is that there would be a MultiBindable which contains two child bindables. It will then dispatch its method calls to BOTH of the children, and take the first result that succeeds.

Reply:
Oh okay, that's an interesting way to do it. I was thinking that this would be done with just the `MultiBindable` type which would have an associated store that holds on to two lenses and two pieces of state but one list of observers. Then the output of the `Bindable` would be a tuple of the output types for the lenses.

Reply:
I think that approach would require that the two lenses both be on models which are built on the same entity, which is not desirable